x-pack/plugins/maps/common/migrations/references.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// Can not use public Layer classes to extract references since this logic must run in both client and server.
+
+import _ from 'lodash';
+
+export function extractReferences({ attributes, references = [] }) {
+  if (!attributes.layerListJSON) {
+    return { attributes, references };
+  }
+
+  const extractedReferences = [];
+
+  const layerList = JSON.parse(attributes.layerListJSON);
+  layerList.forEach((layer, layerIndex) => {
+
+    // Extract index-pattern references from source descriptor
+    if (_.has(layer, 'sourceDescriptor.indexPatternId')) {
+      const refName = `layer_${layerIndex}_index_pattern`;
```

Comment:
I would shoehorn `source` in here somewhere, as the index pattern is on the source, similar to how we name refs for joins.

e.g. `layer[${layerIndex}].source.index_pattern`

Reply:
consider extracting the code for this and for joins in its own helper function.